### ISSUE TITLE
If take_failed_screenshot fails, the rest of the teardown hooks do not run, breaking transactions

### ISSUE THREAD

**Issue by markiz** [rspec/rspec-rails#2401]
## What Ruby, Rails and RSpec versions are you using?

Ruby version: 2.6.3
Rails version: 6.0.3.4

RSpec 3.9

  - rspec-core 3.9.3
  - rspec-expectations 3.9.3
  - rspec-mocks 3.9.1
  - rspec-rails 4.0.1
  - rspec-support 3.9.4

## Observed behaviour

Sometimes take_failed_screenshot would fail, because driver raised an exception. For us, Ferrum::DeadBrowserError was the most common issue. When this happens, all the consequent teardown hooks do not run, including the transactional fixtures teardown (which would normally roll back the wrapper transaction). That leaves the suite with a dirty db state.

## Expected behaviour

I'm not sure. Probably, after_teardown hooks should run even if before_teardown hooks failed?

## Can you provide an example app?

Alas, no, but here's an example patch that helped us with this issue:

```ruby
module ScreenshotHelperPatch
  def take_failed_screenshot(*)
    super
  rescue => e # Silently ignore errors
  end
end

# And in the rails_helper
config.include ScreenshotHelperPatch, type: :system
```

I guess you can emulate this situation by raising from one of the before_teardown callbacks

**Comment by klyonrad** [comment]
Hmm the [code from rails](https://github.com/rails/rails/blob/d415eb4f6d6bb24b78b968ae28c22bb7e1721285/actionpack/lib/action_dispatch/system_testing/test_helpers/setup_and_teardown.rb#L15)  should actually rescue/ensure the calls of `take_failed_screenshot`. @markiz Could you maybe try out reproducing the issue with a non rspec test in your project and report back?

Nevertheless I have a suspicion. Rails tries to call `super` of `before_teardown` when take_failed_screenshot causes an exception. But rspec-rails [defines that method](https://github.com/rspec/rspec-rails/blob/13a8a01cc07c93db97b5f1cd57229527441cd626/lib/rspec/rails/example/system_example_group.rb#L20) as an empty method. That might override whatever rails is trying to calls in the `ensure` block (I don't understand it). Understanding the ancestor tree that rspec-rails file is too complex for me so I might be totally wrong.

**Comment by markiz** [comment]
@klyonrad it does not **rescue**, that's the issue. It bubbles the exception up the chain.

**Comment by JonRowe** [comment]
Yes but neither does Rails, which arguably owns this behaviour, this must happen in mini test too no/

**Comment by markiz** [comment]
> Yes but neither does Rails, which arguably owns this behaviour, this must happen in mini test too no?

Unfortunately, I don't know. 

My expectation would be that a failure in a teardown makes spec fail but does not prevent other teardown hooks from being called.
